src/cpp/alohalytics.cc

```diff
@@ -46,172 +49,86 @@
-void Stats::OnMessage(const MQMessage& message, size_t dropped_events) {
-  if (dropped_events) {
-    LOG_IF_DEBUG("Warning:", dropped_events, "events were dropped from the queue.");
+void Stats::GzipAndArchiveFileInTheQueue(const std::string & in_file, const std::string & out_archive) {
+  if (unique_client_id_event_.empty()) {
+    LOG_IF_DEBUG("Warning: unique client id is not set in GzipAndArchiveFileInTheQueue.")
+  } else {
+    LOG_IF_DEBUG("Archiving", in_file, "to", out_archive);
   }
-
-  if (message.ForceUpload()) {
-    if (upload_url_.empty()) {
-      LOG_IF_DEBUG("Warning: Can't upload in-memory events because upload url has not been set.");
-      return;
-    }
-    LOG_IF_DEBUG("Forcing statistics uploading.");
-    if (file_storage_queue_) {
-      // Upload all data, including 'current' file.
-      file_storage_queue_->ForceProcessing(true);
-    } else {
-      std::string buffer = unique_client_id_event_;
-      // TODO(AlexZ): thread-safety?
-      TMemoryContainer copy;
-      copy.swap(memory_storage_);
-      for (const auto& evt : copy) {
-        buffer.append(evt);
-      }
-      LOG_IF_DEBUG("Forcing in-memory statistics uploading.");
-      if (!UploadBuffer(upload_url_, std::move(buffer), debug_mode_)) {
-        // If failed, merge events we tried to upload with possible new ones.
-        memory_storage_.splice(memory_storage_.end(), std::move(copy));
-        LOG_IF_DEBUG("Failed to upload in-memory statistics.");
-      }
+  // Append unique installation id in the beginning of each archived file.
+  // If it is empty then all stats data will become completely anonymous.
+  try {
+    std::string buffer(unique_client_id_event_);
+    {
+      std::ifstream fi;
+      fi.exceptions(std::ifstream::failbit | std::ifstream::badbit);
+      fi.open(in_file, std::ifstream::in | std::ifstream::binary);
+      const size_t id_size = unique_client_id_event_.size();
+      const int64_t file_size = FileManager::GetFileSize(in_file);
+      buffer.resize(id_size + static_cast<std::string::size_type>(file_size));
+      fi.read(&buffer[id_size], static_cast<std::streamsize>(file_size));
     }
-  } else {
-    if (file_storage_queue_) {
-      file_storage_queue_->PushMessage(message.GetMessage());
-    } else {
-      auto& container = memory_storage_;
-      container.push_back(message.GetMessage());
-      constexpr size_t kMaxEventsInMemory = 2048;
-      if (container.size() > kMaxEventsInMemory) {
-        container.pop_front();
-        LOG_IF_DEBUG("Warning: maximum numbers of events in memory (", kMaxEventsInMemory,
-                     ") was reached and the oldest one was dropped.");
-      }
+    {
+      std::ofstream fo;
+      fo.exceptions(std::ifstream::failbit | std::ifstream::badbit);
+      fo.open(out_archive, std::ofstream::out | std::ofstream::binary | std::ofstream::trunc);
+      const std::string gzipped_buffer = Gzip(buffer);
+      buffer.resize(0);
```

Comment:
`.clear()`, or a swap trick from Meyers (http://en.wikibooks.org/wiki/More_C%2B%2B_Idioms/Shrink-to-fit), ` std::string().swap(buffer);`.